client/blb/client.go

```diff
@@ -1130,136 +1128,113 @@ func (cli *Client) readOneTractReplicated(
+			if err == core.ErrVersionMismatch {
+				badVersionHost = host
+			}
+			if err != core.NoError && err != core.ErrEOF {
+				// TODO(eric): redo bad TS reporting mechanism.
+				ch <- tractResult{0, 0, err, badVersionHost}
 				return
 			}
+			log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
+			// In case of short read, i.e., read < len(thisB), we should
+			// pad the result with 0's. This doesn't need to be done
+			// explicitly if 'thisB' given to us starts to be empty.
+			// However, just in case it's not, we will ensure that by
+			// overwritting untouched bytes with 0's.
+			for i := read; i < len(thisB); i++ {
+				thisB[i] = 0
+			}
+			ch <- tractResult{len(thisB), read, err, badVersionHost}
+			return
+		}
 
-			// Compute alternate hosts that will receive backup requests. Pass them along
-			// to the tracserver rpc layer to enable cancellation messages to be sent between
-			// servers. TODO(eric): come up with a better name for this.
-			otherHosts := cutSlice(tract.Hosts, j)
-
-			_, badVersionHost, err = cli.readIntoReportBadTS(ctx, curAddr, host, otherHosts, reqID, result, tract, thisB, thisOffset)
-			if err != core.NoError {
-				errc <- err
-				return
+		// Backup request logic involves the following:
+		// 1) Send the first read to the first host (item 0 available from the orderCh) in a goroutine.
+		// 2) Start another goroutine to issue backup requests. Each subsequent request is spawned after the
+		// configured backup time.After delay time expires.
+		// 3) Block until the first read is returned from the goroutines servicing the reads. Note that if the
+		// the request returned fails, we accept that and continume on to the fallback phase for sequential reads.
+		// We could optimize this later and scan for the first successful backup request, but that may not be
+		// worth the latency savings.
+		go func() {
+			nb := min(maxNumBackups+1, len(order))
+			for i := 0; i < nb; i++ {
+				select {
+				case <-cli.backupReadState.backupDelayFunc(time.Duration(i) * delay):
+					go readFunc(order[i])
+				case <-ctx.Done():
+					log.V(1).Infof("backup read %s-%d to tractserver cancelled", tract.Tract, i)
+					break
+				}
 			}
-			ch <- result
+		}()
+
+		// take the first result.
+		*result = <-ch
+		if result.err == core.NoError {
 			return
-		}(i, n)
+		}
 	}
-	// block until one goroutine finishes, or all fail due to errors.
-	nerr := 0
-	for {
-		select {
-		case result = <-ch:
-			return
-		case err = <-errc:
-			nerr++
-			if nerr == len(order) {
-				log.V(1).Infof("read %s all hosts failed", tract.Tract)
-				*result = tractResult{0, 0, err, badVersionHost}
-				return
-			}
+
+	// Sequentailly read from tract server replicas. This phase will run if backups are disabled from config state, or
+	// the backup read phase fails and we use this as a fallback.
+	for n := range order {
+		host := tract.Hosts[n]
```

Comment:
Can you consolidate the body of this for loop with `readFunc`?

Reply:
done.